mvm/machine/signer.go

```diff
@@ -4,76 +4,86 @@ import (
 	"bytes"
 	"context"
 	"encoding/binary"
-	"encoding/hex"
+	"fmt"
+	"sort"
 	"time"
 
 	"github.com/MixinNetwork/mixin/logger"
 	"github.com/MixinNetwork/tip/crypto"
 	"github.com/MixinNetwork/tip/crypto/en256"
+	"github.com/MixinNetwork/trusted-group/mvm/constants"
 	"github.com/MixinNetwork/trusted-group/mvm/encoding"
 	"github.com/drand/kyber/sign/tbls"
 )
 
+func (m *Machine) getProcessInfo(processId string) (string, string, bool) {
+	m.procLock.Lock()
+	defer m.procLock.Unlock()
+	process, ok := m.processes[processId]
+	if !ok {
+		return "", "", false
+	}
+	return process.Platform, process.Address, true
+}
+
+func (m *Machine) GetLastSendTime(id string) time.Time {
+	m.smLock.Lock()
+	defer m.smLock.Unlock()
+	return m.sm[id]
+}
+
+func (m *Machine) SetLastSendTime(id string, t time.Time) {
+	m.smLock.Lock()
+	defer m.smLock.Unlock()
+	m.sm[id] = t
+}
+
 func (m *Machine) loopSignGroupEvents(ctx context.Context) {
-	sm := make(map[string]time.Time)
 	for {
 		time.Sleep(3 * time.Second)
 		events, err := m.store.ListPendingGroupEvents(100)
 		if err != nil {
 			panic(err)
 		}
+
 		for _, e := range events {
-			e.Signature = nil
-			logger.Verbosef("Machine.loopSignGroupEvents() => %v", e)
-			msg := e.Encode()
-			partials, err := m.store.ReadPendingGroupEventSignatures(e.Process, e.Nonce)
-			if err != nil {
-				panic(err)
+			lst := m.GetLastSendTime(e.ID()).Add(time.Minute * 5)
+			if lst.After(time.Now()) {
+				continue
 			}
+			m.SetLastSendTime(e.ID(), time.Now())
+			logger.Verbosef("Machine.loopSignGroupEvents() => %d, %v", e.Nonce, e)
 
-			if checkFullSignature(partials) {
-				e.Signature = partials[0]
-				logger.Verbosef("loopSignGroupEvents() => WriteSignedGroupEventAndExpirePending(%v) full", e)
-				err = m.store.WriteSignedGroupEventAndExpirePending(e)
-				if err != nil {
-					panic(err)
-				}
-				continue
+			var partial []byte
+			msg := e.Encode()
```

Comment:
I think this should just make an assertion that e.Signature is nil?

Reply:
MVM process panic if sign failed in SignEvent.

Reply:
You didn't get what this means to do, e.Signature = nil before e.Encode() to ensure the msg is plain without signature. That's why I'm proposing to just panic when e.Signature is not nil. 

Because I think in this situation, no need to make e.Signature = nil, they are pending, so e.Signature should be just nil, otherwise it's a bug.

e.g. 

```
if e.Signature != nil {
panic(e)
}
msg := e.Encode()
```

And in the code below, no need to do this anymore.